### ISSUE TITLE
Add ability to get the viewport of a camera in world coordinates

### ISSUE THREAD

**Issue by ChristianGreiner** [AdamsLair/duality#681]
### Summary

As the title says it would be nice to get the world viewport coordinates of a camera.

For example it could be useful to figure out if objects are in the viewport of the camera.

**Example:**
![randomcontrol_af3_camera_object](https://user-images.githubusercontent.com/6233308/45381866-d6868800-b606-11e8-8357-e400db5fba4c.png)

**Comment by SirePi** [comment]
Can I haz for next month?
Maybe have it return a `Vector2[]` instead of `Rect` to be able to also take into account situations where the camera is rotated.

**Comment by ilexp** [comment]
> Can I haz for next month?

Sure! Assigned you. This should branch off of `develop-3.0`, as I'm trying to close down `master` as we slowly approach v3 release.

> Maybe have it return a Vector2[] instead of Rect to be able to also take into account situations where the camera is rotated.

That's a very good point. However, a Vector2 array is a lot less useful than a rect in most cases where the camera isn't rotated, and it's also one allocation per call - maybe we can find another way?

There's also another issue I hadn't thought about earlier, which is that the rendered image size is unknown - we can't assume [DualityApp.TargetViewSize](https://github.com/AdamsLair/duality/blob/24c4ce8179b30ee3f4f32148102af7f972749d3c/Source/Core/Duality/DualityApp.cs#L181) in the general case, as anyone calling `Render` manually could just [specify any value](https://github.com/AdamsLair/duality/blob/24c4ce8179b30ee3f4f32148102af7f972749d3c/Source/Core/Duality/Components/Camera.cs#L179). So we should probably expose this as an optional parameter, which falls back to the target view size.

So here's my proposal: Let's change this to 

`public Rect GetWorldViewportBounds(float z, Vector2 imageSize = default(Vector2))`

where the default `imageSize` is interpreted as a fallback to `DualityApp.TargetViewSize`, and the returned rect is the axis-aligned bounding rectangle that incorporates all four corners of a the viewport. In the usual, unrotated case, this will be equal to the actual viewport rect. In the rotated case, we still get useful information from the bounds rect. In cases where people want the actual rotated four corners, they can still call `GetWorldPos` four times.

**Comment by ilexp** [comment]
Actually, doesn't hurt to have both methods. Let's do something like this:

```
public Rect GetWorldViewportBounds(float z, Vector2 imageSize = default(Vector2));
public void GetWorldViewportCorners(Vector2[] corners, float z, Vector2 imageSize = default(Vector2));
```

The bounds variant could use the corners one with an internal, static `Vector2[]` to avoid the allocation, and just do the bounding rectangle around that.

**Comment by SirePi** [comment]
So if I understand correctly, the void one would fill the corners parameter? what if the array is not big enough?

**Comment by ilexp** [comment]
> So if I understand correctly, the void one would fill the corners parameter? 

Yep.

> what if the array is not big enough?

Argument check and throw an exception if null or too small.

**Comment by SirePi** [comment]
Starting to think about this.. would it make sense to also have the z parameter optional, and if not set assume it's `this.GameObj.Transform.Pos.Z + this.FocusDist` where this is the Camera?

**Comment by ilexp** [comment]
> Starting to think about this.. would it make sense to also have the z parameter optional, and if not set assume it's this.GameObj.Transform.Pos.Z + this.FocusDist where this is the Camera?

Hmm.. the default might as well be the Z-0 plane, but I'd say both make a similar amount of sense. I'd say this is ambiguous enough to not have a default / optional z.

**Comment by ilexp** [comment]
Implemented by SirePi in #688. Closing this.
